### PR TITLE
feat: support query-based requests

### LINKED ISSUE #5: signature error with cyberduck
1. set up `rclone serve s3` with accessKey & secretKey

```
➜  ~ rclone --version
rclone v1.67.0
- os/version: darwin 14.5 (64 bit)
- os/kernel: 23.5.0 (arm64)
- os/type: darwin
- os/arch: arm64 (ARMv8 compatible)
- go/version: go1.22.4
- go/linking: dynamic
- go/tags: none
```

```
➜  ~ cat ~/.config/rclone/rclone.conf

[local]
type = local
```

```
➜  ~ rclone serve s3 --auth-key aaaa,ssss local:/
2024/07/06 01:39:51 NOTICE: Local file system at /: Starting s3 server on [http://127.0.0.1:8080/]
```

2. Open Cyberduck, and install the`S3 (HTTP)` profile in Cyberduck

image

3. Configure it (I'm sure I have entered the correct accessKey and secretKey)

image

4. Connect. List bucket works fine
image

5. Listing files in a bucket would fail with signature error

image

---

P.S. using rclone to connect to this S3 server works fine. 

---

cc @ncw

### LINKED ISSUE #7616: serve s3: Allow Presigned Requests for Serving S3
#### The associated forum post URL from `https://forum.rclone.org`

https://forum.rclone.org/t/presigned-keys-in-rclone-serve-s3/44449

#### What is your current rclone version (output from `rclone version`)?

```
rclone v1.65.2
- os/version: ubuntu 20.04 (64 bit)
- os/kernel: 5.4.0-33-generic (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.21.6
- go/linking: static
- go/tags: none
```

#### What problem are you are trying to solve?

Allowing Presigned Keys to be used when serving S3. Using the S3 API you can generate a pre-signed URL just fine, but when it comes to actually using that pre-signed url, it fails with 403 Not Authorized Errors, disguised as a 400 Bad Request (Rclone logs it as 403, but responds with 400 on the url itself).

#### How do you think rclone should be changed to solve that?

Implementing authentication verification using the presigned URL. When a user goes to this type of link, it should check authentication from the AWSAcessKeyID and the Signature to see if it matches the Access Key and Secret Key (I don't know how it is calculated or verified, but I don't think S3 keeps track of it manually, I think it is some kind of hashing or something). I hope it isn't stored, as I have multiple S3 serves running to the same data located around the world for faster downloads, so it shouldn't just be stored on the server the presigned key request was sent to.

#### Link Example (not real)

https://storage.example.app/bucket/file.mkv?AWSAccessKeyId=vHNhZ9FZK124XXXXXXXXXXXXXwme0UDfSQAJdSDXXXXX0lduUqsh&Signature=Wb6YUQEMXXXXXXXXY8%3D&Expires=1707179008

#### Documentation about Presigned Keys
- https://boto3.amazonaws.com/v1/documentation/api/latest/guide/s3-presigned-urls.html
- https://docs.aws.amazon.com/AmazonS3/latest/userguide/ShareObjectPreSignedURL.html
- https://github.com/minio/minio/blob/7aa00bff898e6a44f452f6dba802af8d0d198374/cmd/signature-v2.go#L109
- https://github.com/minio/minio/blob/7aa00bff898e6a44f452f6dba802af8d0d198374/cmd/signature-v4.go#L207
- https://github.com/search?q=repo%3Aminio%2Fminio%20presign&type=code

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

#### Bounty

I am also willing to put up a bounty to get this done as fast as possible as I use it for my business. Let me know how that process is done, or if it is even possible at all. Thanks!

### PR DESCRIPTION
**Pull request by itsHenry35**
fix https://github.com/rclone/gofakes3/issues/5
fix https://github.com/rclone/rclone/issues/7616

### CODE REVIEW

**Comment by KevinWang15** [comment]
Hi @itsHenry35,

Thank you for your timely pull request.

I've tested your code, and I noticed that [this change](https://github.com/rclone/gofakes3/pull/6/files#diff-5306d5b0f8c88b148ffa772f859fed982963272c2fe89c9f1e18b544075afdcfL160-R174) might not always be correct.



In most cases, `amzDate` is present in the header.

As an alternative, we could consider the following approach:

```go
var date string
if date = req.Header.Get(amzDate); date == "" {
    if date = r.Header.Get(headerDate); date == "" {
        if date = queryf.Get(amzDate); date == "" {
            return errMissingDateHeader
        }
    }
}
```

Additionally, I'd like to suggest squashing your commits to remove the "[chore: remove generated binary](https://github.com/rclone/gofakes3/pull/6/commits/c2cadeb0f5db83d04d75701c3cb6696ec7f47c6e)" commit before submitting this PR (or you could use `git commit --amend`).

Once again, thanks for the excellent work!

**Comment by itsHenry35** [comment]
> Hi @itsHenry35,
> 
> Thank you for your timely pull request.
> 
> I've tested your code, and I noticed that [this change](https://github.com/rclone/gofakes3/pull/6/files#diff-5306d5b0f8c88b148ffa772f859fed982963272c2fe89c9f1e18b544075afdcfL160-R174) might not always be correct.
> 
> 
> In most cases, `amzDate` is present in the header.
> 
> As an alternative, we could consider the following approach:
> 
> ```go
> var date string
> if date = req.Header.Get(amzDate); date == "" {
>     if date = r.Header.Get(headerDate); date == "" {
>         if date = queryf.Get(amzDate); date == "" {
>             return errMissingDateHeader
>         }
>     }
> }
> ```
> 
> Additionally, I'd like to suggest squashing your commits to remove the "[chore: remove generated binary](https://github.com/rclone/gofakes3/pull/6/commits/c2cadeb0f5db83d04d75701c3cb6696ec7f47c6e)" commit before submitting this PR (or you could use `git commit --amend`).
> 
> Once again, thanks for the excellent work!

Oh, i was opposed to just add a if judgement instead of replacing it, but I did it by mistake. Sorry for the fault, I will correct it

**Comment by KevinWang15** [comment]
@itsHenry35 Great, I have tested your code. Now it works. LGTM

cc @ncw PTAL

File: signature/signature-v4.go
Comment (ncw) [inline signature/signature-v4.go]: This will never be true. What is this test for?

File: signature/signature-v4.go
Comment (ncw) [inline signature/signature-v4.go]: I'd prefer these on two lines as at first glance it appears you've shadowed `date` using a very common go idiom but you haven't as there is an `=` instead of a `:=`

```go
date = queryf.Get(amzDate)
if date == "" {
```

File: signature/signature-v4.go
Comment (ncw) [inline signature/signature-v4.go]: Can't we detect the user is using "UNSIGNED-PAYLOAD" by looking in `X-Amz-Content-Sha256`? Then we don't need to guess .

**Review by ncw — COMMENTED** [review]
Thank you for doing this - every step towards greater compatibility is a good one :-)

I put some comments inline for discussion.

Thanks

Nick

**Edit** we should also think of some tests!

File: signature/signature-v4.go
Comment (itsHenry35) [reply signature/signature-v4.go]: https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html
Please have a look as a reference. When the client generates a web url for easier accessing on web browsers, the authorization headers are passed in query-mode instead of headers. In this case, the v4Auth will be "" and we can get the  auth params from the url query.

File: signature/signature-v4.go
Comment (itsHenry35) [reply signature/signature-v4.go]: Thanks for the suggestion. That sounds reasonable, I'll do some further tests.

File: signature/signature-v4.go
Comment (KevinWang15) [reply signature/signature-v4.go]: This code change is for support for "pre-signed URLs", a feature quite commonly used in S3.

Example (https://runkit.com/kevinwang15/668e1f70270f8e0008046ad9):

```javascript
const AWS = require("aws-sdk");

const s3 = new AWS.S3({
    endpoint: 'https://127.0.0.1',
    accessKeyId: 'aaaa',
    secretAccessKey: 'ssss',
    s3ForcePathStyle: true,
    signatureVersion: 'v4',
    sslEnabled: true,
});


function generateDownloadUrl(bucket, key) {
    const params = {
        Bucket: bucket,
        Key: key,
        Expires: 3600,
    };

    s3.getSignedUrl('getObject', params, (err, url) => {
        if (err) {
            console.error(err);
        } else {
            console.log(url);
        }
    });
}

generateDownloadUrl('bucket1', 'a.txt');

```

It produces URLs like

> https://127.0.0.1/bucket1/a.txt?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=aaaa%2F20240710%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240710T053620Z&X-Amz-Expires=3600&X-Amz-Signature=623b5d21aa2698af14d046d40008a72ef4ba3dfed9bb90d412e6dba4a6ec8706&X-Amz-SignedHeaders=host

This URL can be used anywhere - send a GET request and it will give you the file.

**Comment by KevinWang15** [comment]
Code has been updated, LGTM
@ncw PTAL again, thanks

File: signature/signature-v4.go
Comment (ncw) [reply signature/signature-v4.go]: I think we are talking at cross purposes! I agree with what you said above but it wasn't my question.

My specific query is this:

In this line we set `v4Auth` to something

```go
v4Auth = fmt.Sprintf("%s Credential=%s, SignedHeaders=%s, Signature=%s", queryf.Get(amzAlgorithm), queryf.Get(amzCredential), queryf.Get(amzSignedHeaders), queryf.Get(amzSignature))
```

This can never be `""`

And in the line after we check it against `""` which can never be true surely?

```go
if v4Auth == "" {
```

Which is indicative of some kind of logic error.

Comment (itsHenry35) [reply signature/signature-v4.go]: Uh sorry, it's my fault, I'm going to correct it.

Comment (itsHenry35) [reply signature/signature-v4.go]: Thanks for correction. I've corrected this logical error, pls have a look at the new force-pushed code.

**Review by ncw — APPROVED** [review]
This is looking great now - thank you :-)

I'll just wait for the CI to go green then I'll merge.

Do you want to propose a PR for rclone to update the go.mod to fix the corresponding rclone issue?

**Comment by itsHenry35** [comment]
> This is looking great now - thank you :-)
> 
> I'll just wait for the CI to go green then I'll merge.
> 
> Do you want to propose a PR for rclone to update the go.mod to fix the corresponding rclone issue?

Yeah, of course. I will propose the PR after the merge. Thx ^:^

**Comment by ncw** [comment]
Thank you - great job :-)

**Comment by itsHenry35** [comment]
> This is looking great now - thank you :-)
> 
> I'll just wait for the CI to go green then I'll merge.
> 
> Do you want to propose a PR for rclone to update the go.mod to fix the corresponding rclone issue?

@ncw I've submitted the PR. Feel free to review it. Thanks! https://github.com/rclone/rclone/pull/7944

**Pull request merged**
